### PR TITLE
chore(db): Propagate errors to the client.

### PR DESCRIPTION
**Pull request by calavera**
This change propagates write errors to the application using SlateDB.

I've removed the calls to `expect` that cause the process to panic on error, and wrapped a few missing errors into SlateDbError. I've also unified `MemtableFlushThreadMsg` and `WallFlushThreadMsg` into a single structure because the looked redundant.

@criccomini the only panic that I was not sure about is this code below introduced in https://github.com/slatedb/slatedb/pull/264/files#diff-51529babecd6e4018b78a49622df5b5d2c480476389ec07c0b4943d8bdbaa35eR71. To me, it looks like that code is never going to be reachable because the `if` clause always returns true if the feature flag `wal_disable` is not enabled. It makes me wonder if this feature flag is even useful for people. It doesn't look like it adds any additional dependencies or code. I would propose removing the feature flag all together, but maybe it's a discussion for an issue instead.

```
let current_table = if self.wal_enabled() {
...
} else {
            if cfg!(not(feature = "wal_disable")) {
                panic!("wal_disabled feature must be enabled");
            }
```

I think when this is merged, #111 can be closed.

### CODE REVIEW

**Comment by criccomini** [comment]
Looks good!

> To me, it looks like that code is never going to be reachable because the if clause always returns true if the feature flag wal_disable is not enabled

The `wal_disabled` stuff is purely for paranoia. Disabling the WAL can be dangerous, so @rodesai added a feature flag to make sure users _really_ wanted it. I agree that the if statement in the batch write guarantees that the panic will never occur, but for paranoia's sake I want to keep it. Just in case we ever refactor and miss something.

> I think when this is merged, https://github.com/slatedb/slatedb/issues/111 can be closed.

Your PR is definitely an improvement over where we are, but I don't think it fully resolves #111. The issue I see is that the errors in the WAL flush thread and memtable flush thread never get propagated to the client.

In the WAL flusher, we ignore flush errors when we tick:

https://github.com/slatedb/slatedb/blob/979852c302a3ea494ca934b85d6930d246d66c66/src/flush.rs#L94-L97

And simply print to `error!` when we hit errors in the mem table flush tick:

https://github.com/slatedb/slatedb/blob/979852c302a3ea494ca934b85d6930d246d66c66/src/mem_table_flush.rs#L104-L116

In #111, IIRC, I think we discussed adding an `error` field to `db_state` (https://github.com/slatedb/slatedb/issues/111#issuecomment-2322945882) and checking it from the client operations. I think this is still required at least for the two flush threads I mentioned. I suspect we need something similar for the compactor and GC threads as well.

**Review by criccomini — APPROVED** [review]

File: src/flush.rs [hunk 1]
```diff
@@ -11,9 +11,9 @@ use crate::iter::KeyValueIterator;
 use crate::mem_table::{ImmutableWal, KVTable, WritableKVTable};
 use crate::types::ValueDeletable;
 
-pub(crate) enum WalFlushThreadMsg {
```
[the remark this answers is not in the document]
Comment (criccomini) [reply src/flush.rs:14]: SGTM. @calavera  mind fixing in a new PR?

[the remark this answers is not in the document]
Comment (calavera) [reply src/flush.rs:14]: yes, I'll put this back in a new PR.

[the remark this answers is not in the document]
Comment (calavera) [reply src/flush.rs:14]: https://github.com/slatedb/slatedb/pull/304

**Pull request merged**